File: server/routers/admin/src/product_management/attributes/schema.rs
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;
use validator::Validate;

#[derive(Debug, ToSchema, Validate, Deserialize, Serialize)]
pub struct AttributeValue {
    #[validate(
        length(min = 1, message = "Value is required"),
        length(max = 50, message = "Maximum 50 characters are allowed")
    )]
    #[schema(min_length = 1, max_length = 50)]
    pub value: String,
}

#[derive(Debug, ToSchema, Validate, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateAttributeInput {
    #[validate(
        length(min = 1, message = "Name is required"),
        length(max = 50, message = "Maximum 50 characters are allowed")
    )]
    #[schema(example = "Color", min_length = 1, max_length = 50)]
    /// Name of the attribute.
    ///
    pub name: String,

    #[validate]
    /// Values of the attribute.
    pub values: Vec<AttributeValue>,
}

File: client/apps/admin/src/app/(dashboard)/product-management/attributes/_lib/schema.ts
import { type CreateAttributeInput } from "@/client";
import { string, withSchema } from "@repo/utils/zod";
import { array, object } from "zod";

export const createAttributeSchema = withSchema<CreateAttributeInput>()({
  name: string()
    .min(1, "Name is required")
    .max(50, "Maximum 50 characters are allowed"),
  values: array(
    object({
      value: string()
        .min(1, "Value is required")
        .max(50, "Maximum 50 characters are allowed"),
    }),
  ),
});
